File: src/spectrl/context.py
"""Compact selected-spectrum context and namespaced extension contracts."""

from __future__ import annotations

import copy
import re

from .model import SpectrlCvParam, SpectrlUserParam

FIELDS = {
    "params": 0,
    "user_params": 1,
    "id": 2,
    "name": 3,
    "version": 4,
    "location": 5,
    "external_ids": 6,
    "spectrum_ref": 7,
    "instrument": 8,
    "components": 9,
    "kind": 10,
    "order": 11,
    "software": 12,
    "operation": 13,
    "revision": 14,
    "parameters": 15,
    "source_params": 16,
}
ALLOWED = {
    "source": {"params", "user_params", "id", "name", "location", "external_ids", "spectrum_ref"},
    "acquisition": {"instrument"},
    "instrument": {"params", "user_params", "id", "name", "components", "software"},
    "component": {"params", "user_params", "kind", "order"},
    "software": {"params", "user_params", "id", "name", "version"},
    "processing": {"params", "user_params", "software", "operation", "revision", "parameters", "source_params"},
}
EXTENSIONS = {}
_NAMESPACE = re.compile(r"^[A-Za-z][A-Za-z0-9._-]*:[A-Za-z0-9._/-]+$")


def _params(values, cls):
    if not isinstance(values, list):
        raise ValueError("parameters must be a list")
    return [cls(**x) if isinstance(x, dict) else x for x in values]


def encode_record(value, kind):
    from .header import _encode_param_map, _encode_user_params

    if not isinstance(value, dict) or set(value) - ALLOWED[kind]:
        raise ValueError(f"invalid {kind} record fields")
    out = {}
    for key, val in value.items():
        if val is None:
            continue
        if key in {"params", "source_params"}:
            val = _encode_param_map(_params(val, SpectrlCvParam))
        elif key == "user_params":
            val = _encode_user_params(_params(val, SpectrlUserParam))
        elif key in {"instrument", "software"}:
            val = encode_record(val, key)
        elif key == "components":
            if not isinstance(val, list):
                raise ValueError("components must be a list")
            val = [encode_record(x, "component") for x in val]
        elif key == "external_ids":
            if not isinstance(val, list) or not all(isinstance(x, str) and x for x in val):
                raise ValueError("external_ids must be a list of nonempty strings")
        elif key in {"order", "revision"}:
            if type(val) is not int or not (0 if key == "order" else 1) <= val <= 9007199254740991:
                raise ValueError(f"{key} must be a valid nonnegative order or positive revision")
        elif key == "parameters":
            if not isinstance(val, dict) or not all(isinstance(x, str) for x in val):
                raise ValueError("processing parameters must be a string-keyed map")
        elif not isinstance(val, str):
            raise ValueError(f"{key} must be a string")
        if key == "kind" and val not in {"source", "analyzer", "detector"}:
            raise ValueError("invalid instrument component kind")
        out[FIELDS[key]] = val
    return out


def decode_record(value, kind):
    from .header import _decode_param_map, _decode_user_params

    reverse = {v: k for k, v in FIELDS.items()}
    if not isinstance(value, dict) or any(type(k) is not int or k not in reverse for k in value):
        raise ValueError(f"invalid {kind} record")
    out = {}
    for raw_key, val in value.items():
        key = reverse[raw_key]
        if key in {"params", "source_params"}:
            val = _decode_param_map(val)
        elif key == "user_params":
            val = _decode_user_params(val)
        elif key in {"instrument", "software"}:
            val = decode_record(val, key)
        elif key == "components":
            if not isinstance(val, list):
                raise ValueError("components must be a list")
            val = [decode_record(x, "component") for x in val]
        out[key] = val
    encode_record(out, kind)
    return out


def register_extension(identifier, validator, *, revision=1):
    if not isinstance(identifier, str) or not _NAMESPACE.fullmatch(identifier):
        raise ValueError("extension identifier must be namespaced")
    if type(revision) is not int or revision < 1 or not callable(validator):
        raise ValueError("invalid extension registration")
    if (identifier, revision) in EXTENSIONS:
        raise ValueError("extension already registered")
    EXTENSIONS[identifier, revision] = validator


def validate_extensions(value, *, require_supported=True):
    if not isinstance(value, dict):
        raise ValueError("extensions must be a map")
    for identifier, record in value.items():
        if not isinstance(identifier, str) or not _NAMESPACE.fullmatch(identifier):
            raise ValueError("extension identifier must be namespaced")
        if not isinstance(record, dict) or set(record) != {"revision", "required", "data"}:
            raise ValueError("extension requires revision, required, and data")
        revision = record["revision"]
        if type(revision) is not int or not 1 <= revision <= 9007199254740991 or type(record["required"]) is not bool:
            raise ValueError("invalid extension revision or required flag")
        validator = EXTENSIONS.get((identifier, revision))
        if validator:
            validator(record["data"])
        elif require_supported and record["required"]:
            raise ValueError(f"unsupported required extension {identifier}@{revision}")


def check_array_mutation(spec):
    """Unknown extension dependencies cannot safely survive a permutation or selection."""
    if spec.extensions or any(spec.array_extensions.values()):
        raise ValueError("array mutation requires explicitly removing or updating extensions first")


def record_change(spec, operation, parameters):
    """Keep acquisition-derived summaries under an explicit source scope."""
    summary_terms = {"MS:1000285", "MS:1000504", "MS:1000505", "MS:1000527", "MS:1000528"}
    summaries = [p for p in spec.params if p.accession in summary_terms]
    event = {"operation": operation, "revision": 1, "parameters": parameters}
    if summaries:
        event["source_params"] = copy.deepcopy(summaries)
    return [p for p in spec.params if p.accession not in summary_terms], [*spec.processing, event]

File: js/src/context.ts
/** Selected-spectrum context and namespaced extension contracts. */
import type { ContextRecord, Extensions, InlineSpectrum } from "./model.js"
import { encodeParamMap, decodeParamMap, encodeUserParams, decodeUserParams, type MsgMap } from "./header.js"
export const fields = ["params", "userParams", "id", "name", "version", "location", "externalIds", "spectrumRef", "instrument", "components", "kind", "order", "software", "operation", "revision", "parameters", "sourceParams"] as const
const allowed: Record<string, string[]> = {
  source: ["params", "userParams", "id", "name", "location", "externalIds", "spectrumRef"], acquisition: ["instrument"],
  instrument: ["params", "userParams", "id", "name", "components", "software"], component: ["params", "userParams", "kind", "order"],
  software: ["params", "userParams", "id", "name", "version"], processing: ["params", "userParams", "software", "operation", "revision", "parameters", "sourceParams"],
}
export function toWire(value: unknown): unknown {
  if (value instanceof Map) return new Map([...value].map(([k, v]) => [k, toWire(v)]))
  if (Array.isArray(value)) return value.map(toWire)
  if (value && typeof value === "object" && !(value instanceof Uint8Array)) return new Map(Object.entries(value).map(([k, v]) => [k, toWire(v)]))
  return value
}
export function fromWire(value: unknown): any {
  if (value instanceof Map) {
    if ([...value.keys()].some(k => typeof k !== "string")) return new Map([...value].map(([k, v]) => [k, fromWire(v)]))
    return Object.fromEntries([...value].map(([k, v]) => [k, fromWire(v)]))
  }
  if (Array.isArray(value)) return value.map(fromWire)
  return value
}
/** A plain object with string keys: not null, an array, a Map, a typed array,
 * or any other exotic object that merely reports `typeof "object"`. */
function isStringKeyedObject(value: unknown): value is Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value) || value instanceof Map || ArrayBuffer.isView(value)) return false
  const proto = Object.getPrototypeOf(value)
  if (proto !== Object.prototype && proto !== null) return false
  return Reflect.ownKeys(value).every(k => typeof k === "string")
}

export function encodeRecord(value: ContextRecord, kind: string): MsgMap {
  if (!value || typeof value !== "object" || Array.isArray(value) || Object.keys(value).some(k => !allowed[kind]!.includes(k))) throw Error(`invalid ${kind} record`)
  const out: MsgMap = new Map()
  for (const [key, val] of Object.entries(value)) {
    if (val == null) continue
    let encoded: unknown = val
    if (key === "params" || key === "sourceParams") encoded = encodeParamMap(val)
    else if (key === "userParams") encoded = encodeUserParams(val)
    else if (key === "instrument" || key === "software") encoded = encodeRecord(val, key)
    else if (key === "components") {
      if (!Array.isArray(val)) throw Error("components must be an array")
      encoded = val.map(x => encodeRecord(x, "component"))
    } else if (key === "externalIds") {
      if (!Array.isArray(val) || val.some(x => typeof x !== "string" || !x.length)) throw Error("externalIds require nonempty strings")
    } else if (key === "order" || key === "revision") {
      if (!Number.isSafeInteger(val) || val < (key === "order" ? 0 : 1)) throw Error("order must be nonnegative and revision positive")
    } else if (key === "parameters") {
      // A byte string is `typeof "object"` and is neither an Array nor a Map,
      // so it slipped through an earlier shape check and was accepted as a
      // parameter map. Require a plain object with string keys, as section 7 does.
      if (!isStringKeyedObject(val)) throw Error("parameters require a string-keyed object")
      encoded = toWire(val)
    } else if (typeof val !== "string") throw Error(`${key} must be a string`)
    if (key === "kind" && !["source", "analyzer", "detector"].includes(val)) throw Error("invalid component kind")
    out.set(fields.indexOf(key as typeof fields[number]), encoded)
  }
  return out
}
export function decodeRecord(value: unknown, kind: string): ContextRecord {
  if (!(value instanceof Map)) throw Error(`invalid ${kind} record`)
  const out: Record<string, unknown> = {}
  for (const [k, v] of value) {
    if (!Number.isInteger(k) || k < 0 || k >= fields.length) throw Error(`invalid ${kind} field`)
    const key = fields[k]!
    let decoded = v
    if (key === "params" || key === "sourceParams") decoded = decodeParamMap(v)
    else if (key === "userParams") decoded = decodeUserParams(v)
    else if (key === "instrument" || key === "software") decoded = decodeRecord(v, key)
    else if (key === "components") {
      if (!Array.isArray(v)) throw Error("components must be an array")
      decoded = v.map(x => decodeRecord(x, "component"))
    } else if (key === "parameters") decoded = fromWire(v)
    out[key] = decoded
  }
  encodeRecord(out, kind)
  return out
}
const namespace = /^[A-Za-z][A-Za-z0-9._-]*:[A-Za-z0-9._/-]+$/
const extensions = new Map<string, (data: unknown) => void>()
export function registerExtension(id: string, validate: (data: unknown) => void, revision = 1) {
  if (!namespace.test(id) || !Number.isSafeInteger(revision) || revision < 1 || typeof validate !== "function") throw Error("invalid extension registration")
  const key = JSON.stringify([id, revision])
  if (extensions.has(key)) throw Error("extension already registered")
  extensions.set(key, validate)
}
export function validateExtensions(value: Extensions, requireSupported = true) {
  if (!value || typeof value !== "object" || Array.isArray(value) || value instanceof Map) throw Error("extensions must be a map")
  for (const [id, r] of Object.entries(value)) {
    if (!namespace.test(id) || !r || typeof r !== "object" || Object.keys(r).sort().join() !== "data,required,revision" || !Number.isSafeInteger(r.revision) || r.revision < 1 || typeof r.required !== "boolean") throw Error("invalid extension record")
    const validate = extensions.get(JSON.stringify([id, r.revision]))
    if (validate) validate(r.data)
    else if (r.required && requireSupported) throw Error(`unsupported required extension ${id}@${r.revision}`)
  }
}
export function checkArrayMutation(spec: InlineSpectrum) {
  if (Object.keys(spec.extensions ?? {}).length || Object.values(spec.arrayExtensions ?? {}).some(x => Object.keys(x).length)) throw Error("array mutation requires explicitly removing or updating extensions first")
}
export function recordChange(spec: InlineSpectrum, operation: string, parameters: Record<string, unknown>) {
  const summaries = new Set(["MS:1000285", "MS:1000504", "MS:1000505", "MS:1000527", "MS:1000528"])
  const sourceParams = (spec.params ?? []).filter(p => summaries.has(p.accession))
  return { params: (spec.params ?? []).filter(p => !summaries.has(p.accession)), processing: [...spec.processing ?? [], { operation, revision: 1, parameters, ...(sourceParams.length ? { sourceParams } : {}) }] }
}
export function withoutUserParams(spec: InlineSpectrum): InlineSpectrum {
  let count = 0
  const strip = (value: any, opaque = false): any => {
    if (opaque || value == null || typeof value !== "object" || ArrayBuffer.isView(value)) return value
    if (Array.isArray(value)) return value.map(v => strip(v))
    return Object.fromEntries(Object.entries(value).map(([k, v]) => {
      if (k === "userParams") { count += (v as unknown[]).length
        return [k, []]
      }
      if (k === "arrayUserParams") { count += Object.values(v as Record<string, unknown[]>).reduce((n, xs) => n + xs.length, 0)
        return [k, {}]
      }
      return [k, strip(v, ["extensions", "arrayExtensions", "parameters"].includes(k))]
    }))
  }
  const out = strip(spec) as InlineSpectrum
  if (count) out.processing = [...out.processing ?? [], { operation: "spectrl:metadata-omission", revision: 1, parameters: { userParamsRemoved: count } }]
  return out
}
